### PR TITLE
BUGFIX: Check if array keys exist before accessing them

### PR DESCRIPTION
**Pull request by p-weisk**
See Issue #27

### CODE REVIEW

**Comment by Sebobo** [comment]
Thanks for the PR!

I think it would make sense to store the checked variables in a separate variable instead of checking them in so many places. 
The many conditions based on the `label` could also be compressed into a smarter one I think. But that's something I can do if you want. Would make the code nicer.

**Comment by p-weisk** [comment]
Ok, I'll refactor it.

**Comment by p-weisk** [comment]
I refactored it a bit to make it more legible overall, but I could use some help with Testing since I do not have access to a matomo instance which returns actual values into those variables (browser type, device type, os, outlink).

File: Classes/Domain/Dto/BrowserDataResult.php [hunk 1]
```diff
@@ -22,12 +22,11 @@ function jsonSerialize()
         foreach ($this->results as $year => $devices) {
             if (is_array($devices)) {
                 foreach ($devices as $device) {
-                    $totalVisits = $totalVisits + $device['nb_visits'];
```
Comment (Sebobo) [inline Classes/Domain/Dto/BrowserDataResult.php]: You first have to calculate `total Visits` as it's used in in the percentage calculation. 
Without that the numbers wouldn't be right anymore. 
Of course array_reduce would be better than the foreach.

File: Classes/Domain/Dto/DeviceDataResult.php [hunk 2]
```diff
@@ -27,16 +27,12 @@ function jsonSerialize()
         foreach ($this->results as $year => $devices) {
             if (is_array($devices)) {
                 foreach ($devices as $device) {
-                    if ($device['label'] == 'Desktop') {
-                        $clientDevices['Desktop'] = $clientDevices['Desktop'] + $device['nb_visits'];
+                    $label = $device['label'] ?? 0;
```
Comment (Sebobo) [inline Classes/Domain/Dto/DeviceDataResult.php]: As label is expected to be a string, the fallback should be one of `Desktop`, `Tablet` or  `Smartphone` and I tend to the first.

File: Classes/Domain/Dto/OperatingSystemDataResult.php [hunk 3]
```diff
@@ -30,22 +30,12 @@ function jsonSerialize()
         foreach ($this->results as $year => $devices) {
             if (is_array($devices)) {
                 foreach ($devices as $device) {
-                    if ($device['label'] == 'GNU/Linux') {
-                        $clientOperatingSystems['GNU/Linux'] = $clientOperatingSystems['GNU/Linux'] + $device['nb_visits'];
+                    $label = $device['label'] ?? 0;
```
Comment (Sebobo) [inline Classes/Domain/Dto/OperatingSystemDataResult.php]: Also here label should be a string matching one of the actual OS or an empty string.

File: Classes/Domain/Dto/OutlinkDataResult.php [hunk 4]
```diff
@@ -23,13 +23,11 @@ function jsonSerialize()
         foreach ($this->results as $year => $devices) {
             if (is_array($devices)) {
                 foreach ($devices as $device) {
-                    $totalVisits = $totalVisits + $device['nb_hits'];
```
Comment (Sebobo) [inline Classes/Domain/Dto/OutlinkDataResult.php]: TotalVisits need to be calculated first, so you can't combine the two loops yet.

**Comment by p-weisk** [comment]
@Sebobo thanks for the review and the "type hint" for the labels, I resolved your comments, please review again.

**Pull request merged**